src/main/java/com/github/sdonkov/PrefixTree.java

```diff
@@ -3,16 +3,50 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.util.AbstractCollection;
 import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
 import java.util.Optional;
 
-public class PrefixTree {
+public class PrefixTree extends AbstractCollection<String> {
```

Comment:
The move to String is unnecessary. Read "ITEM 31: USE BOUNDED WILDCARDS TO INCREASE API FLEXIBILITY" from the book and adjust accordingly. You'll have to make the class a generic one.

I'd suggest going for a clean implementation of the Collection interface instead. It'll bring you more knowledge which is the final goal here; otherwise, I agree this would save some boilerplate code, but not that much.